ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt

```diff
@@ -209,4 +218,67 @@ class SingleSignOnProfileVerifier(private val authnRequest: AuthnRequest,
         }
         return audienceRestriction.first()
     }
+
+    /**
+     * 3.1 Holder of Key
+     */
+    internal fun verifyHolderOfKey() {
+        val holderOfKeyList = samlResponseDom.recursiveChildren(SUBJECT_CONFIRMATION)
+            .filter { it.attributeText("Method") == HOLDER_OF_KEY_URI }
+
+        if (holderOfKeyList.isEmpty()) return
+
+        val holderOfKeyDataList = holderOfKeyList
+            .flatMap { it.children(SUBJECT_CONFIRMATION_DATA) }
+
+        if (holderOfKeyDataList.isEmpty())
+            throw SAMLComplianceException.create(SAMLProfiles_3_1_a,
+                message = "<SubjectConfirmationData> not found within Holder of Key " +
+                    "<SubjectConfirmation>",
+                node = samlResponseDom)
+
+        holderOfKeyDataList.forEach {
+            verifyHolderOfKeyData(it)
+        }
+    }
+
+    private fun verifyHolderOfKeyData(node: Node) {
+        val type = node.attributeNodeNS(XSI, "type")
+        if (type != null) {
+            val (namespace, value) = type.textContent.split(":")
+            if (value != KEY_INFO_CONFIRMATION_DATA_TYPE)
+                throw SAMLComplianceException.createWithPropertyMessage(SAMLProfiles_3_1_b,
+                    property = "type",
+                    actual = value,
+                    expected = KEY_INFO_CONFIRMATION_DATA_TYPE,
+                    node = node)
+
+            // SSO Response must have at least one assertion with an assertion namespace
```

Comment:
`type?.firstChild?.namespaceURI` does not return the namespace uri for the value. Couldn't find a way to get the namespace of a value from dom so decided to compare it to the prefix of assertions.
In SSO Assertions are required and have the namespace we need.

Reply:
That returns null as well 😞

Reply:
:/